### DISCUSSION TITLE
Possible regression in 1.3.0?

### DISCUSSION THREAD

**Discussion by kmnhan** [lmfit/lmfit-py#953] [Q&A] [answered]
I was exploiting the behavior in https://github.com/lmfit/lmfit-py/blob/cf2626b7de95ae7271b1bf0352fa76015a40f126/lmfit/model.py#L516-L520 to create a function with variable number of input parameters, by passing in a callable class with `argnames` and `kwargs` properties.
For a minimal example, I could make a polynomial function with an arbitrary degree with the following code:

```python
import lmfit
import numpy as np


class PolynomialFunction:
    def __init__(self, degree: int = 1):
        self.degree = degree

    @property
    def __name__(self) -> str:
        return self.__class__.__name__

    @property
    def argnames(self) -> list[str]:
        return ["x"] + [f"c{i}" for i in range(self.degree + 1)]

    @property
    def kwargs(self):
        return {}

    def __call__(self, x, *coeffs, **params):
        if len(coeffs) != self.degree + 1:
            coeffs = [params[f"c{d}"] for d in range(self.degree + 1)]
        return np.polynomial.polynomial.polyval(x, coeffs)


model = lmfit.Model(PolynomialFunction(degree=2))
model.make_params()
```

This broke with 1.3.0 due to changes made in #941 because `default_vals` is not defined for the first `if` block. It looks like a simple fix, but I am unsure whether this is supported public API. Will this be removed entirely in the future? If so, how should I write a function/model that accomplishes the same thing?

**Comment by newville** [comment, marked as answer]
@kmnhan Ah, sorry for the trouble.  Well, I would say both that
   a) we never really promised this was part of a supported API that would never change ;)
   b) but also: variadic functions are an important-if-maybe-challenging use case, and we want them to work!

As you noticed, the changes for 1.3.0 from #941 were a fairly large change to the internals of parsing and converting function arguments to parameters and independent variables.  I do not expect this to be something we change very often, but we may have to make some tweaks from those changes to fix bugs.  See #950, for example.   

I think we could (and should) probably fix the code so your use case works too.  I guess we weren't testing that portion of the code (or at least not very thoroughly!).   We should be able to fold fixes for this into #950, and then push that out as 1.3.1 fairly soon. 

But also: we do have a `PolynomialModel` builtin(`lmfit.models.PolynomialModel`), and also `lmfit.models.SplineModel` that are both variadic.  Would it be reasonable to use that builtin? if not, are there changes we should make?

**Reply by kmnhan** [reply]
Thank you for the response! I will keep an eye out for the 1.3.1 release. 
I've been using this approach for more complex models, but looking at the implementation of the builtin models, I think I can start refactoring my models so that they are less.. hacky. Thanks for the suggestion!
